apollo-router/src/metrics/mod.rs

```diff
@@ -0,0 +1,48 @@
+use crate::metrics::aggregation::AggregateMeterProvider;
+use once_cell::sync::OnceCell;
+
+pub(crate) mod aggregation;
+pub(crate) mod filter;
+pub(crate) mod layer;
+
+static AGGREGATE_METER_PROVIDER: OnceCell<AggregateMeterProvider> = OnceCell::new();
+pub(crate) fn meter_provider() -> &'static AggregateMeterProvider {
+    AGGREGATE_METER_PROVIDER.get_or_init(Default::default)
+}
+
+#[macro_export]
+macro_rules! metric_counter_u64 {
+    ($name:literal, $description:literal, $value: expr, $($attr_key:expr => $attr_value:expr),+) => {
+        let meter_provider = crate::metrics::meter_provider();
+        let meter = opentelemetry::metrics::MeterProvider::meter(meter_provider, "apollo/meter");
+        let counter = meter.u64_counter($name).with_description($description).init();
+        let attributes = Vec::new();
+        $(attributes.push(KeyValue::new($attr_key, $attr_value));)+
+        counter.add($value, &attributes);
```

Comment:
this will probably have a huge cost at runtime. tracing's macro is less flexible, but it had the advantage of creating a static structure that reduced the need for allocations. See https://docs.rs/tracing/latest/src/tracing/macros.rs.html#587-598 and https://github.com/tokio-rs/tracing/blob/b8c45ccd067a6aebdd9257c5edda6bc90bcf9028/tracing/src/macros.rs#L2764-L2772
It could make sense to have such a structure for static fields, and a separate part for more dynamic fields
Could you profile the router and see how those metrics behave?

Reply:
The plan is to add some caching, but I'm just getting things correct first. Hold off on reviewing just yet.

Reply:
Sorry, but this is exactly the right time to discuss this, because caching has the potential to change the behaviour. And when working on something that can affect performance, it's often easier to test if the solution would be fast and reliable enough, then spend the time to make it correct. What design do you have in mind?

Reply:
One idea is that callsite will hold a static mutex<weak> of the actual instrument returned by the aggregate meter provider. If the weak reference cannot be upgraded then a new intstrument is fetched.

This way there are zero map lookups when once a callsite has been populated, it goes directly to the instrument. 

Maybe the instead of an mutex it's an rwlock, but as the duration of the lock is the upgrade of a weak then it's debatable if it's worth it.

Considering that the metrics layer had to do a bunch of string manipulation on *every* call I think the above will likely be faster.

Reply:
However I reserve the right to completely change my mind on the above. Like I said, I'm focusing on correctness first, I appreciate people work in different ways, but this is the way that I prefer to approach problems.

Reply:
Resolving this as callsite caching has now been implemented and we can start a new perf conversation fresh.